### PR TITLE
test: adding BTT tests for BNum (cont. #122)

### PR DESCRIPTION
**Pull request by wei3erHase**
Closes BAL-133

Continuation of #122 but targetting `dev` branch.

### CODE REVIEW

File: test/unit/BNum.tree [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+├── when passing a value less than BONE
+│   └── it should return zero
+└── when using known values
+    └── it should return correct value
+
+
+BNumTest::badd
+├── when passing zero and zero
+│   └── it should return zero
+├── when passing a as uint256 max and b non zero
```
Comment (0xteddybear) [inline test/unit/BNum.tree:24]: nit: use left & right when both arguments would have the same name?

File: test/unit/BNum.tree [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+├── when passing b as uint256 max and a non zero
+│   └── it should revert
+└── when passing known values
+    └── it should return correct value
+
+
+BNumTest::bsub
+├── when passing zero and zero
+│   └── it should return zero
+├── when passing a less than b
```
Comment (0xteddybear) [inline test/unit/BNum.tree:35]: nit: use minuend and sustraend or at least left/right?

File: test/unit/BNum.tree [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+│   └── it should revert
+├── when passing a mul b too big
+│   └── it should revert
+│       └── a * b + BONE / 2 > uint256 max
+└── when passing known values
+    └── it should return correct value
+
+
+BNumTest::bdiv
+├── when passing b as zero
```
Comment (0xteddybear) [inline test/unit/BNum.tree:65]: missing case where a=0?

File: test/unit/BNum.t.sol [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+  }
+
+  function test_BdivRevertWhen_PassingAAsUint256Max() external {
+    // it should revert
+    vm.expectRevert(BNum.BNum_DivInternal.selector);
+
+    bNum.call_bdiv(type(uint256).max, 1);
+  }
+
+  function test_BdivWhenPassingAAndBTooBig() external {
```
Comment (0xteddybear) [inline test/unit/BNum.t.sol:233]: this doesn't implement A and B 'too big', A is max (definetely too big) but b is just BONE

File: test/unit/BNum.t.sol [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    uint256 _a = 4e18;
+    uint256 _b = 12;
+
+    uint256 _result = bNum.call_bpowi(_a, _b);
+
+    assertEq(_result, 16_777_216e18);
+  }
+
+  function test_BpowWhenPassingExponentAsZero() external {
+    uint256 _result = bNum.call_bpow(BONE, 0);
```
Comment (0xteddybear) [inline test/unit/BNum.t.sol:309]: opinion on fuzzing the base?

File: test/unit/BNum.tree [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+├── when passing a and b too big
+│   └── it should revert
+│       └── a*BONE/b + b/2 > uint256 max
+├── when flooring to zero
+│   └── it should return 0
+└── when passing known values
+    └── it should return correct value
+
+
+BNumTest::bpowi
```
Comment (0xteddybear) [inline test/unit/BNum.tree:78]: opinion on having a revert when a number with BONE precision is passed as the exponent? other functions already have 'reverts on overflows caused by parameters' covered

**Review by 0xteddybear — DISMISSED** [review]
sorry, I only have weapons-grade nitpicks for you today

File: test/unit/BNum.t.sol (hunk 4 above)
Comment (wei3erHase) [reply test/unit/BNum.t.sol:233]: corrected it as there are 2 errors in BNum that we wanna reach, the 1st one is when `a * BONE` overflows (soon to be renamed to left / right?), and the 2nd one is the one we wanna catch here, that has a `b` crippled (bc is `/2`) and the sum needs to overflow, interesting puzzle

File: test/unit/BNum.tree (hunk 6 above)
Comment (wei3erHase) [reply test/unit/BNum.tree:74]: hmmm i don't think that's actually a branch, but rather a result. gave it a try, and appears that even `(BONE + 1 )^ BONE` is actually possible (being `BONE+1` the minimum base to have an effect) to calculate, but `(BONE + 1)^ 100 * BONE` (arbitrary range) does always overflow (according to foundry fuzzer).

i believe this is more task of echidna than BTT

**Comment by wei3erHase** [comment]
deprecated the formula line after `should revert` because bulloak interpreted it as a regular test, instead of as a `BdivRevertWhen` case

also re-reviewed the tree and applied some tweaks, i think it still covering all the branches, and gave a try to the fuzzer to get all the ranges of overflowing combination of variables

as for `a` vs `b`, i do think they're helpful for when commenting the formulas:
`a + b > uint256 max`
vs
`left + right > uint256 max` (consider it can escalate to a complex one)

i would be OK with using `left` and `right` in the tree (that ends up being `WhenLeftGreaterThanX`), but keep using `a` and `b` for formulas (which may get quite confusing), wdyt?

**Comment by wei3erHase** [comment]
for flooring, i'm adding both a fuzzed and a funny example, the rest i think they're fully fuzzed, didn't end up very complicated, and we're using bounds instead of assumptions, which is still great for speed of execution

File: test/unit/BNum.t.sol [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    // it should revert
+    //      a*BONE + b/2 > uint256 max
+    vm.expectRevert(BNum.BNum_DivInternal.selector);
+
+    bNum.call_bdiv(_a, _b);
+  }
+
+  function test_BdivWhenFlooringToZero() external {
+    // it should return zero
+    //     (1 * BONE) / 2 * BONE + 1 = 0.499..
```
Comment (0xteddybear) [inline test/unit/BNum.t.sol:242]: `(1*BONE)/(2*BONE+1)` ?

**Review by 0xteddybear — APPROVED** [review]
good work! 

All in on doing 1 test with hardcoded values an another one with fuzzed inputs

**Review by 0xteddybear — APPROVED** [review]
:rocket:

**Pull request merged**